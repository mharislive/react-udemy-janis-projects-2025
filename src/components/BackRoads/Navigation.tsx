type NavigationProps = React.HTMLAttributes<HTMLElement> & {};

const Navigation = ({
  className = "",
  "aria-label": ariaLabel = "main navigation",
}: NavigationProps) => {
  // better approach would be to use refs but it would be too much work for this demo app
  const handleScroll = (
    e: React.MouseEvent<HTMLAnchorElement>,
    section: string
  ) => {
    e.preventDefault();
    const el = document.getElementById(`${section}-section`);
    if (el) {
      el.scrollIntoView({ behavior: "smooth" });
    }
  };

  return (
    <nav
      className={`flex gap-3 justify-center ${className}`}
      aria-label={ariaLabel}
    >
      <a
        href="#"
        className="transition duration-500 hover:text-cyan-500 tracking-widest"
        onClick={(e) => handleScroll(e, "hero")}
      >
        Home
      </a>
      <a
        href="#"
        className="transition duration-500 hover:text-cyan-500 tracking-widest"
        onClick={(e) => handleScroll(e, "about")}
      >
        About
      </a>
      <a
        href="#"
        className="transition duration-500 hover:text-cyan-500 tracking-widest"
        onClick={(e) => handleScroll(e, "services")}
      >
        Services
      </a>
      <a
        href="#"
        className="transition duration-500 hover:text-cyan-500 tracking-widest"
        onClick={(e) => handleScroll(e, "tours")}
      >
        Tours
      </a>
    </nav>
  );
};

export default Navigation;
