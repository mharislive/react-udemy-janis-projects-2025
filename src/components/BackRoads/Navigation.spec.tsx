import { render, screen, fireEvent } from "@testing-library/react";
import { axe } from "vitest-axe";
import Navigation from "./Navigation";

describe("Navigation", () => {
  it("renders all navigation links", () => {
    render(<Navigation />);
    expect(screen.getByText("Home")).toBeInTheDocument();
    expect(screen.getByText("About")).toBeInTheDocument();
    expect(screen.getByText("Services")).toBeInTheDocument();
    expect(screen.getByText("Tours")).toBeInTheDocument();
  });

  it("applies custom className", () => {
    render(<Navigation className="test" />);
    const nav = screen.getByRole("navigation");
    expect(nav).toHaveClass("test");
  });

  it("calls scrollIntoView on click with correct section", () => {
    const scrollIntoViewMock = vi.fn();
    const sectionId = "about-section";
    const section = document.createElement("div");
    section.id = sectionId;
    section.scrollIntoView = scrollIntoViewMock;
    document.body.appendChild(section);

    render(<Navigation />);
    const aboutLink = screen.getByText("About");
    fireEvent.click(aboutLink);

    expect(scrollIntoViewMock).toHaveBeenCalledWith({ behavior: "smooth" });

    document.body.removeChild(section);
  });

  it("prevents default link behavior", () => {
    render(<Navigation />);
    const homeLink = screen.getByText("Home");
    const event = new MouseEvent("click", {
      bubbles: true,
      cancelable: true,
    });
    Object.defineProperty(event, "target", { value: homeLink });
    const preventDefault = vi.fn();
    homeLink.addEventListener("click", (e) => {
      e.preventDefault = preventDefault;
    });
    homeLink.dispatchEvent(event);
    expect(preventDefault).toHaveBeenCalled();
  });

  it("does not throw if section does not exist", () => {
    render(<Navigation />);
    const servicesLink = screen.getByText("Services");
    expect(() => fireEvent.click(servicesLink)).not.toThrow();
  });

  it("should match snapshot", () => {
    const { asFragment } = render(<Navigation />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should have no accessibility violations", async () => {
    const { container } = render(<Navigation />);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});
