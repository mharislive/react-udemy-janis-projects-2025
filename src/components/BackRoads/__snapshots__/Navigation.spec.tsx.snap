// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Navigation > should match snapshot 1`] = `
<DocumentFragment>
  <nav
    aria-label="main navigation"
    class="flex gap-3 justify-center "
  >
    <a
      class="transition duration-500 hover:text-cyan-500 tracking-widest"
      href="#"
    >
      Home
    </a>
    <a
      class="transition duration-500 hover:text-cyan-500 tracking-widest"
      href="#"
    >
      About
    </a>
    <a
      class="transition duration-500 hover:text-cyan-500 tracking-widest"
      href="#"
    >
      Services
    </a>
    <a
      class="transition duration-500 hover:text-cyan-500 tracking-widest"
      href="#"
    >
      Tours
    </a>
  </nav>
</DocumentFragment>
`;
